File: python/transpose/transpose.py
"""Transpose a block of text."""

def transpose(lines: str) -> str:
    """Transpose a string.

    First, lines need to be space padded such that every line is as long, or longer,
    than the following line. Input:
        "1"
        "333"
        "22"
    becomes:
        "1  "
        "333"
        "22"
    Next, convert each line into a list as long as the first line, padding with "",
    to aid in matrix transposing (to avoid any IndexErrors).
    """
    # Input with space padding so line length never increases.
    space_padded: list[str] = []
    width = 0
    for row in reversed(lines.splitlines()):
        width = max(len(row), width)
        space_padded.append(row.ljust(width))
    space_padded.reverse()

    # Build a "matrix" which is rectangular to aid in lookups.
    input_matrix: list[list[str]] = []
    for row in space_padded:
        input_matrix.append(list(row) + [""] * (width - len(row)))

    # Transpose the data which now has the same number of elements on every line.
    return "\n".join(
        "".join(input_matrix[y][x] for y in range(len(input_matrix)))
        for x in range(width)
    )

File: go/transpose/transpose.go
package transpose

import (
	"slices"
	"strings"
)

// Transpose a slice of strings.
func Transpose(input []string) []string {
	// Compute the longest line length.
	var longest int
	for _, line := range input {
		if len(line) > longest {
			longest = len(line)
		}
	}

	var out []string
	var priorLen int
	for idx := range longest {
		// Work in reverse to handle length padding - each line must be as long as the next line.
		idx = longest - idx - 1
		result := ""
		for _, line := range input {
			if idx < len(line) {
				result += string(line[idx])
			} else {
				result += " "
			}
		}
		// Strip trailing spaces then add back as much as is needed.
		result = strings.TrimRight(result, " ")
		for i := len(result); i < priorLen; i++ {
			result += " "
		}
		priorLen = len(result)
		out = append(out, result)
	}
	slices.Reverse(out)
	return out
}
